src/leap/mail/adaptors/soledad.py

```diff
@@ -0,0 +1,1047 @@
+                    "Cannot create: mdoc has a doc_id")
+        leap_assert(self.fdoc.doc_id is None,
+                    "Cannot create: fdoc has a doc_id")
+
+        # TODO check that the doc_ids in the mdoc are coherent
+        # TODO I think we need to tolerate the no hdoc.doc_id case, for when we
+        # are doing a copy to another mailbox.
+        # leap_assert(self.hdoc.doc_id is None,
+        # "Cannot create: hdoc has a doc_id")
+        d = []
+        d.append(self.mdoc.create(store))
+        d.append(self.fdoc.create(store))
+        if self.hdoc.doc_id is None:
+            d.append(self.hdoc.create(store))
+        for cdoc in self.cdocs.values():
+            if cdoc.doc_id is not None:
+                # we could be just linking to an existing
+                # content-doc.
+                continue
+            d.append(cdoc.create(store))
+        return defer.gatherResults(d)
+
+    def update(self, store):
+        """
+        Update the only mutable parts, which are within the flags document.
+        """
+        return self.fdoc.update(store)
+
+    def delete(self, store):
+        # TODO
+        # Eventually this would have to do the duplicate search or send for the
+        # garbage collector. At least mdoc and t the mdoc and fdoc can be
+        # unlinked.
+        d = []
+        if self.mdoc.doc_id:
+            d.append(self.mdoc.delete(store))
+        d.append(self.fdoc.delete(store))
+        return defer.gatherResults(d)
+
+    def copy(self, store, newmailbox):
+        """
+        Return a copy of this MessageWrapper in a new mailbox.
+        """
+        # 1. copy the fdoc, mdoc
+        # 2. remove the doc_id of that fdoc
+        # 3. create it (with new doc_id)
+        # 4. return new wrapper (new meta too!)
+        raise NotImplementedError()
+
+    def set_mbox_uuid(self, mbox_uuid):
+        """
+        Set the mailbox for this wrapper.
+        This method should only be used before the Documents for the
+        MessageWrapper have been created, will raise otherwise.
+        """
+        mbox_uuid = mbox_uuid.replace('-', '_')
+        self.mdoc.set_mbox_uuid(mbox_uuid)
+        self.fdoc.set_mbox_uuid(mbox_uuid)
+
+    def set_flags(self, flags):
+        # TODO serialize the get + update
+        if flags is None:
+            flags = tuple()
+        leap_assert_type(flags, tuple)
+        self.fdoc.flags = list(flags)
+        self.fdoc.deleted = "\\Deleted" in flags
+        self.fdoc.seen = "\\Seen" in flags
+        self.fdoc.recent = "\\Recent" in flags
+
+    def set_tags(self, tags):
+        # TODO serialize the get + update
+        if tags is None:
+            tags = tuple()
+        leap_assert_type(tags, tuple)
+        self.fdoc.tags = list(tags)
+
+    def set_date(self, date):
+        # XXX assert valid date format
+        self.hdoc.date = date
+
+    def get_subpart_dict(self, index):
+        """
+        :param index: index, 1-indexed
```

Comment:
Now I understand that `1-indexed` means that the indexes start from 1. I'm not sure if it should be more verbose, but at least I think the `index, ` is making it more confusing and should be removed.